x-pack/plugins/endpoint/server/routes/resolver/utils/pagination.ts

```diff
@@ -56,36 +63,45 @@ export function getPaginationParams(limit: number, after?: string): PaginationPa
   return { size: limit };
 }
 
-export function paginate(pagination: PaginationParams, field: string, query: JsonObject) {
+export function paginate(
+  pagination: PaginationParams,
+  tiebreaker: string,
+  aggregator: string,
+  query: JsonObject
+) {
   const { size, timestamp, eventID } = pagination;
-  query.sort = [{ '@timestamp': 'asc' }, { [field]: 'asc' }];
-  query.aggs = { total: { value_count: { field } } };
+  query.sort = [{ '@timestamp': 'asc' }, { [tiebreaker]: 'asc' }];
+  query.aggs = query.aggs || {};
+  query.aggs = Object.assign({}, query.aggs, { totals: { terms: { field: aggregator, size } } });
   query.size = size;
   if (timestamp && eventID) {
     query.search_after = [timestamp, eventID] as Array<number | string>;
   }
   return query;
 }
 
-export function paginatedResults(
-  response: SearchResponse<ResolverEvent>
-): { total: number; results: ResolverEvent[]; nextCursor: string | null } {
-  const total = response.aggregations?.total?.value || 0;
-  if (response.hits.hits.length === 0) {
-    return { total, results: [], nextCursor: null };
+export function buildPaginationCursor(total: number, results: ResolverEvent[]): string | null {
+  if (total > results.length && results.length > 0) {
+    const lastResult = results[results.length - 1];
+    const cursor = {
+      timestamp: lastResult['@timestamp'],
+      eventID: extractEventID(lastResult),
+    };
+    return urlEncodeCursor(cursor);
   }
+  return null;
+}
 
-  const results: ResolverEvent[] = [];
-  for (const hit of response.hits.hits) {
-    results.push(hit._source);
+export function paginatedResults(response: SearchResponse<ResolverEvent>): PaginatedResults {
```

Comment:
I don't understand the naming here, this looks more like it's extracting and reformatting the results than paginating them.